module/zfs/vdev_disk.c

```diff
@@ -83,16 +84,25 @@ vdev_bdev_mode(int smode)
 #endif /* HAVE_OPEN_BDEV_EXCLUSIVE */
 
 static uint64_t
-bdev_capacity(struct block_device *bdev)
+bdev_capacity(struct block_device *bdev, boolean_t wholedisk)
 {
-	struct hd_struct *part = bdev->bd_part;
-
-	/* The partition capacity referenced by the block device */
-	if (part)
+	if (wholedisk) {
+		/* Sectors used by the EFI partition and partition alignment */
+		uint64_t used = EFI_MIN_RESV_SIZE +
+		    NEW_START_BLOCK +
+		    PARTITION_END_ALIGNMENT;
+		/* Sectors available to the vdev */
+		uint64_t available = get_capacity(bdev->bd_disk) - used;
+
+		/* Bytes available to the vdev */
+		return (available << 9);
+	} else {
+		struct hd_struct *part = bdev->bd_part;
+		ASSERT(part);
+
+		/* The partition capacity referenced by the block device */
 		return (part->nr_sects << 9);
-
-	/* Otherwise assume the full device capacity */
-	return (get_capacity(bdev->bd_disk) << 9);
+	}
```

Comment:
Disks can also be partition-less, typically seen in multi-path configurations. Not sure if we need a third case here?

Reply:
I think it makes sense to have a third case, at least to match the behavior of the previous implementation. I'm thinking something like this:
```
    if (part)
        if (wholedev) 
            // corrected capacity
        else:
            //partsize
    else:
       get_capacity(bdev->bd_disk)
````